pkg/controller/iam/serviceaccountkey.go

```diff
@@ -0,0 +1,336 @@
+
+		return managed.ExternalCreation{
+			ExternalNameAssigned: true,
+			ConnectionDetails:    connDetails,
+		}, nil
+	}
+}
+
+func (s *serviceAccountKeyExternalClient) Update(_ context.Context,
+	_ resource.Managed) (managed.ExternalUpdate, error) {
+	// ServiceAccountKeys are immutable, i.e.,GCP IAM Rest API does not provide an update method:
+	// https://cloud.google.com/iam/docs/reference/rest/v1/projects.serviceAccounts.keys
+	return managed.ExternalUpdate{}, nil
+}
+
+func (s *serviceAccountKeyExternalClient) Delete(ctx context.Context, mg resource.Managed) error {
+	cr, ok := mg.(*v1alpha1.ServiceAccountKey)
+
+	if !ok {
+		return errors.New(errNotServiceAccountKey)
+	}
+
+	rrn, extNameExists, err := s.rrn.resourcePath(cr)
+
+	if err != nil {
+		return errors.Wrap(err, errDeleteServiceAccountKey)
+	}
+
+	if !extNameExists {
+		return nil
+	}
+
+	_, err = s.serviceAccountKeyClient.Delete(rrn).Context(ctx).Do()
+
+	if gcp.IsErrorNotFound(err) {
+		return nil
+	}
+
+	return errors.Wrap(err, errDeleteServiceAccountKey)
+}
+
+// NewRelativeResourceNamer makes an instance of the RelativeResourceNamer
+// which is the only type that is allowed to know how to construct GCP resource names
+// for the IAM type.
+func newSaKeyRelativeResourceNamer(projectID string) saKeyrelativeResourceNamer {
+	return saKeyrelativeResourceNamer{
+		saNamer: RelativeResourceNamer{
+			projectName: projectID,
+		},
+	}
+}
+
+// saKeyrelativeResourceNamer allows the controller to generate the Google Cloud API relative resource name
+// for the service account key based on the external-name annotation.
+// https://cloud.google.com/apis/design/resource_names#relative_resource_name
+// The relative resource names for service account keys have the following format:
+// projects/{project_id}/serviceAccounts/{account}/keys/{key}
+type saKeyrelativeResourceNamer struct {
+	saNamer RelativeResourceNamer
+}
+
+// referencedServiceAccountPath returns the external name of the service account which saKey belongs to
+func (rrn saKeyrelativeResourceNamer) referencedServiceAccountPath(saKey *v1alpha1.ServiceAccountKey) (string, error) {
+	// we assume Google Cloud API relative resource name (aka resource path) of the ServiceAccount is stored in spec.ForProvider.ServiceAccount
+	if saKey.Spec.ForProvider.ServiceAccount == nil || *saKey.Spec.ForProvider.ServiceAccount == "" {
+		return "", fmt.Errorf(fmtErrInvalidServiceAccountRef, saKey.Spec.ForProvider.ServiceAccountReferer)
+	}
+
+	return *saKey.Spec.ForProvider.ServiceAccount, nil
+}
+
+// resourcePath yields the Google Cloud API relative resource name for the ServiceAccountKey resource
+//   returns <the relative resource name>, <whether the external name annotation is non-empty>, <error encountered during resolution>
+func (rrn saKeyrelativeResourceNamer) resourcePath(saKey *v1alpha1.ServiceAccountKey) (string, bool, error) {
+	if saPath, err := rrn.referencedServiceAccountPath(saKey); err != nil {
```

Comment:
This check might be redundant since crossplane-runtime managed reconciler ensures non of observe/create/update/delete called before references got resolved: https://github.com/crossplane/crossplane-runtime/blob/f2418dc9cdf42e9224ccbdaa6f1fd1b1f46782c2/pkg/reconciler/managed/reconciler.go#L588-L614

Reply:
We could remove the check which asserts a valid `.spec.forProvider.ServiceAccount` in `saKeyrelativeResourceNamer.referencedServiceAccountPath` here:

```
	if saKey.Spec.ForProvider.ServiceAccount == nil || *saKey.Spec.ForProvider.ServiceAccount == "" {
		return "", fmt.Errorf(fmtErrInvalidServiceAccountRef, saKey.Spec.ForProvider.ServiceAccountReferer)
	}
```

given that managed reconciler will make sure it's filled. But a violation of this assumption (because of a future change, etc.) will result in a panic. So I had preferred an abundance of caution here.

Reply:
The `ServiceAccount` is actually a required information for this resource but it might be resolved in runtime so we can't mark it as required in CRD. However, I think making its type `string` instead of `*string` and relying on that would be OK. If the requests fail due to missing service account name, GCP will let us know anyway.

Reply:
I don't want to enter a lengthy discussion here but I have a few concerns:
1. Because the referred  ServiceAccount is resolved after the referrer is admitted, my understanding is that we have to keep it as optional. Here, a mutating admission controller seems like a good solution (I'd like to work on something like this if possible :).
2. K8s API conventions ask optional fields to be declared as pointer types unless the field has a built-in nil value, even if the zero-value for the field implies that the field is not set.
3. The referrer API (v1alpha1.ServiceAccountReferer) is shared with already published resources.